{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Шестое домашнее задание. Оценка качества и сборка различных технологий\n",
    "\n",
    "<div style=\"text-align: right\">\n",
    "    Чангалиди Антон\n",
    "    </div>\n",
    "\n",
    "\n",
    "1. Сборка длинных ридов\n",
    "\n",
    "Соберите данные `PacBio` с разным покрытием любым понравившимся вам сборщиком, например, [`Canu`](https://github.com/marbl/canu/releases), [`Flye`](https://github.com/fenderglass/Flye) или [`minasm`](https://github.com/lh3/miniasm/releases). Учтите, что сборка последнего нуждается также в коррекции ошибок. \n",
    "\n",
    "Выберу Flye:\n",
    "\n",
    "Установка: `conda install -c bioconda flye`\n",
    "\n",
    "https://github.com/fenderglass/Flye/blob/flye/docs/USAGE.md#quickusage\n",
    "\n",
    "Написано, что необходимо знать примерную длину генома, и организм. Из референсного файла я получаю информацию:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4706042 data/reference.fasta\n",
      ">gi|48994873|gb|U00096.2| Escherichia coli str. K-12 substr. MG1655, complete genome\n"
     ]
    }
   ],
   "source": [
    "!wc -m data/reference.fasta\n",
    "!cat data/reference.fasta | grep -e '>'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.7M, это эшерихия коли."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2020-04-07 11:53:49] INFO: Starting Flye 2.7-b1585\n",
      "[2020-04-07 11:53:49] INFO: >>>STAGE: configure\n",
      "[2020-04-07 11:53:49] INFO: Configuring run\n",
      "[2020-04-07 11:53:49] INFO: Total read length: 54515681\n",
      "[2020-04-07 11:53:49] INFO: Input genome size: 4700000\n",
      "[2020-04-07 11:53:49] INFO: Estimated coverage: 11\n",
      "[2020-04-07 11:53:49] INFO: Reads N50/N90: 7475 / 2952\n",
      "[2020-04-07 11:53:49] INFO: Minimum overlap set to 3000\n",
      "[2020-04-07 11:53:49] INFO: Selected k-mer size: 15\n",
      "[2020-04-07 11:53:49] INFO: >>>STAGE: assembly\n",
      "[2020-04-07 11:53:49] INFO: Assembling disjointigs\n",
      "[2020-04-07 11:53:49] INFO: Reading sequences\n",
      "[2020-04-07 11:53:50] INFO: Generating solid k-mer index\n",
      "[2020-04-07 11:54:05] INFO: Counting k-mers (1/2):\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 11:54:07] INFO: Counting k-mers (2/2):\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 11:54:11] INFO: Filling index table\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 11:54:19] INFO: Extending reads\n",
      "[2020-04-07 11:54:22] INFO: Overlap-based coverage: 3\n",
      "[2020-04-07 11:54:22] INFO: Median overlap divergence: 0.184516\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 11:54:29] INFO: Assembled 125 disjointigs\n",
      "[2020-04-07 11:54:29] INFO: Generating sequence\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 11:54:54] INFO: >>>STAGE: consensus\n",
      "[2020-04-07 11:54:54] INFO: Running Minimap2\n",
      "[2020-04-07 11:55:07] INFO: Computing consensus\n",
      "[2020-04-07 11:55:27] INFO: Alignment error rate: 0.191619\n",
      "[2020-04-07 11:55:27] INFO: >>>STAGE: repeat\n",
      "[2020-04-07 11:55:27] INFO: Building and resolving repeat graph\n",
      "[2020-04-07 11:55:27] INFO: Parsing disjointigs\n",
      "[2020-04-07 11:55:27] INFO: Building repeat graph\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 11:55:48] INFO: Median overlap divergence: 0.0855838\n",
      "[2020-04-07 11:55:48] INFO: Parsing reads\n",
      "[2020-04-07 11:55:48] INFO: Aligning reads to the graph\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 11:56:12] INFO: Aligned read sequence: 33022822 / 48855793 (0.675924)\n",
      "[2020-04-07 11:56:12] INFO: Median overlap divergence: 0.118117\n",
      "[2020-04-07 11:56:12] INFO: Mean edge coverage: 7\n",
      "[2020-04-07 11:56:12] INFO: Simplifying the graph\n",
      "[2020-04-07 11:56:12] INFO: >>>STAGE: contigger\n",
      "[2020-04-07 11:56:12] INFO: Generating contigs\n",
      "[2020-04-07 11:56:12] INFO: Reading sequences\n",
      "[2020-04-07 11:56:13] INFO: Generated 98 contigs\n",
      "[2020-04-07 11:56:13] INFO: Added 0 scaffold connections\n",
      "[2020-04-07 11:56:13] INFO: >>>STAGE: polishing\n",
      "[2020-04-07 11:56:13] INFO: Polishing genome (1/1)\n",
      "[2020-04-07 11:56:13] INFO: Running minimap2\n",
      "[2020-04-07 11:56:23] INFO: Separating alignment into bubbles\n",
      "[2020-04-07 11:56:42] INFO: Alignment error rate: 0.139108\n",
      "[2020-04-07 11:56:42] INFO: Correcting bubbles\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:01:58] INFO: >>>STAGE: finalize\n",
      "[2020-04-07 12:01:58] INFO: Assembly statistics:\n",
      "\n",
      "\tTotal length:\t4336541\n",
      "\tFragments:\t97\n",
      "\tFragments N50:\t60526\n",
      "\tLargest frg:\t155636\n",
      "\tScaffolds:\t0\n",
      "\tMean coverage:\t9\n",
      "\n",
      "[2020-04-07 12:01:58] INFO: Final assembly: /home/toharhymes/work/ngs_ib_spring2020/project_6_evaluaton/assembling_cov10/assembly.fasta\n"
     ]
    }
   ],
   "source": [
    "!flye --pacbio-raw data/pacbio_10x.fq --out-dir assembling_cov10 --threads 8 -g 4.7m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2020-04-07 12:01:59] INFO: Starting Flye 2.7-b1585\n",
      "[2020-04-07 12:01:59] INFO: >>>STAGE: configure\n",
      "[2020-04-07 12:01:59] INFO: Configuring run\n",
      "[2020-04-07 12:01:59] INFO: Total read length: 109142883\n",
      "[2020-04-07 12:01:59] INFO: Input genome size: 4700000\n",
      "[2020-04-07 12:01:59] INFO: Estimated coverage: 23\n",
      "[2020-04-07 12:01:59] INFO: Reads N50/N90: 7563 / 2948\n",
      "[2020-04-07 12:01:59] INFO: Minimum overlap set to 3000\n",
      "[2020-04-07 12:01:59] INFO: Selected k-mer size: 15\n",
      "[2020-04-07 12:01:59] INFO: >>>STAGE: assembly\n",
      "[2020-04-07 12:01:59] INFO: Assembling disjointigs\n",
      "[2020-04-07 12:01:59] INFO: Reading sequences\n",
      "[2020-04-07 12:02:00] INFO: Generating solid k-mer index\n",
      "[2020-04-07 12:02:16] INFO: Counting k-mers (1/2):\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:02:21] INFO: Counting k-mers (2/2):\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:02:32] INFO: Filling index table\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:02:44] INFO: Extending reads\n",
      "[2020-04-07 12:02:48] INFO: Overlap-based coverage: 7\n",
      "[2020-04-07 12:02:48] INFO: Median overlap divergence: 0.185354\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:02:58] INFO: Assembled 74 disjointigs\n",
      "[2020-04-07 12:02:58] INFO: Generating sequence\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:03:41] INFO: >>>STAGE: consensus\n",
      "[2020-04-07 12:03:41] INFO: Running Minimap2\n",
      "[2020-04-07 12:04:16] INFO: Computing consensus\n",
      "[2020-04-07 12:05:00] INFO: Alignment error rate: 0.194345\n",
      "[2020-04-07 12:05:00] INFO: >>>STAGE: repeat\n",
      "[2020-04-07 12:05:00] INFO: Building and resolving repeat graph\n",
      "[2020-04-07 12:05:00] INFO: Parsing disjointigs\n",
      "[2020-04-07 12:05:00] INFO: Building repeat graph\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:05:25] INFO: Median overlap divergence: 0.0455536\n",
      "[2020-04-07 12:05:25] INFO: Parsing reads\n",
      "[2020-04-07 12:05:25] INFO: Aligning reads to the graph\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:06:21] INFO: Aligned read sequence: 71599487 / 97811235 (0.732017)\n",
      "[2020-04-07 12:06:21] INFO: Median overlap divergence: 0.117335\n",
      "[2020-04-07 12:06:21] INFO: Mean edge coverage: 14\n",
      "[2020-04-07 12:06:21] INFO: Simplifying the graph\n",
      "[2020-04-07 12:06:21] INFO: >>>STAGE: contigger\n",
      "[2020-04-07 12:06:21] INFO: Generating contigs\n",
      "[2020-04-07 12:06:21] INFO: Reading sequences\n",
      "[2020-04-07 12:06:22] INFO: Generated 26 contigs\n",
      "[2020-04-07 12:06:22] INFO: Added 0 scaffold connections\n",
      "[2020-04-07 12:06:22] INFO: >>>STAGE: polishing\n",
      "[2020-04-07 12:06:22] INFO: Polishing genome (1/1)\n",
      "[2020-04-07 12:06:22] INFO: Running minimap2\n",
      "[2020-04-07 12:06:56] INFO: Separating alignment into bubbles\n",
      "[2020-04-07 12:07:45] INFO: Alignment error rate: 0.137944\n",
      "[2020-04-07 12:07:45] INFO: Correcting bubbles\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:11:06] INFO: >>>STAGE: finalize\n",
      "[2020-04-07 12:11:07] INFO: Assembly statistics:\n",
      "\n",
      "\tTotal length:\t4683515\n",
      "\tFragments:\t26\n",
      "\tFragments N50:\t314698\n",
      "\tLargest frg:\t718724\n",
      "\tScaffolds:\t0\n",
      "\tMean coverage:\t19\n",
      "\n",
      "[2020-04-07 12:11:07] INFO: Final assembly: /home/toharhymes/work/ngs_ib_spring2020/project_6_evaluaton/assembling_cov20/assembly.fasta\n"
     ]
    }
   ],
   "source": [
    "!flye --pacbio-raw data/pacbio_20x.fq --out-dir assembling_cov20 --threads 8 -g 4.7m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2020-04-07 12:11:07] INFO: Starting Flye 2.7-b1585\n",
      "[2020-04-07 12:11:07] INFO: >>>STAGE: configure\n",
      "[2020-04-07 12:11:07] INFO: Configuring run\n",
      "[2020-04-07 12:11:09] INFO: Total read length: 218843316\n",
      "[2020-04-07 12:11:09] INFO: Input genome size: 4700000\n",
      "[2020-04-07 12:11:09] INFO: Estimated coverage: 46\n",
      "[2020-04-07 12:11:09] INFO: Reads N50/N90: 7543 / 2957\n",
      "[2020-04-07 12:11:09] INFO: Minimum overlap set to 3000\n",
      "[2020-04-07 12:11:09] INFO: Selected k-mer size: 15\n",
      "[2020-04-07 12:11:09] INFO: >>>STAGE: assembly\n",
      "[2020-04-07 12:11:09] INFO: Assembling disjointigs\n",
      "[2020-04-07 12:11:09] INFO: Reading sequences\n",
      "[2020-04-07 12:11:10] INFO: Generating solid k-mer index\n",
      "[2020-04-07 12:11:28] INFO: Counting k-mers (1/2):\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:11:38] INFO: Counting k-mers (2/2):\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:11:54] INFO: Filling index table\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:12:27] INFO: Extending reads\n",
      "[2020-04-07 12:12:33] INFO: Overlap-based coverage: 13\n",
      "[2020-04-07 12:12:33] INFO: Median overlap divergence: 0.182674\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:13:03] INFO: Assembled 14 disjointigs\n",
      "[2020-04-07 12:13:03] INFO: Generating sequence\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:14:08] INFO: >>>STAGE: consensus\n",
      "[2020-04-07 12:14:08] INFO: Running Minimap2\n",
      "[2020-04-07 12:15:18] INFO: Computing consensus\n",
      "[2020-04-07 12:16:41] INFO: Alignment error rate: 0.196868\n",
      "[2020-04-07 12:16:41] INFO: >>>STAGE: repeat\n",
      "[2020-04-07 12:16:41] INFO: Building and resolving repeat graph\n",
      "[2020-04-07 12:16:41] INFO: Parsing disjointigs\n",
      "[2020-04-07 12:16:41] INFO: Building repeat graph\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:17:02] INFO: Median overlap divergence: 0.0397424\n",
      "[2020-04-07 12:17:02] INFO: Parsing reads\n",
      "[2020-04-07 12:17:03] INFO: Aligning reads to the graph\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:17:46] INFO: Aligned read sequence: 145297516 / 196329755 (0.740069)\n",
      "[2020-04-07 12:17:46] INFO: Median overlap divergence: 0.114903\n",
      "[2020-04-07 12:17:46] INFO: Mean edge coverage: 29\n",
      "[2020-04-07 12:17:46] INFO: Simplifying the graph\n",
      "[2020-04-07 12:17:46] INFO: >>>STAGE: contigger\n",
      "[2020-04-07 12:17:46] INFO: Generating contigs\n",
      "[2020-04-07 12:17:46] INFO: Reading sequences\n",
      "[2020-04-07 12:17:47] INFO: Generated 1 contigs\n",
      "[2020-04-07 12:17:47] INFO: Added 0 scaffold connections\n",
      "[2020-04-07 12:17:47] INFO: >>>STAGE: polishing\n",
      "[2020-04-07 12:17:47] INFO: Polishing genome (1/1)\n",
      "[2020-04-07 12:17:47] INFO: Running minimap2\n",
      "[2020-04-07 12:18:33] INFO: Separating alignment into bubbles\n",
      "[2020-04-07 12:20:16] INFO: Alignment error rate: 0.136631\n",
      "[2020-04-07 12:20:16] INFO: Correcting bubbles\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:22:36] INFO: >>>STAGE: finalize\n",
      "[2020-04-07 12:22:36] INFO: Assembly statistics:\n",
      "\n",
      "\tTotal length:\t4642057\n",
      "\tFragments:\t1\n",
      "\tFragments N50:\t4642057\n",
      "\tLargest frg:\t4642057\n",
      "\tScaffolds:\t0\n",
      "\tMean coverage:\t39\n",
      "\n",
      "[2020-04-07 12:22:36] INFO: Final assembly: /home/toharhymes/work/ngs_ib_spring2020/project_6_evaluaton/assembling_cov40/assembly.fasta\n"
     ]
    }
   ],
   "source": [
    "!flye --pacbio-raw data/pacbio_40x.fq --out-dir assembling_cov40 --threads 8 -g 4.7m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2020-04-07 12:22:36] INFO: Starting Flye 2.7-b1585\n",
      "[2020-04-07 12:22:36] INFO: >>>STAGE: configure\n",
      "[2020-04-07 12:22:36] INFO: Configuring run\n",
      "[2020-04-07 12:22:39] INFO: Total read length: 436994771\n",
      "[2020-04-07 12:22:39] INFO: Input genome size: 4700000\n",
      "[2020-04-07 12:22:39] INFO: Estimated coverage: 92\n",
      "[2020-04-07 12:22:39] INFO: Reads N50/N90: 7566 / 2967\n",
      "[2020-04-07 12:22:39] INFO: Minimum overlap set to 3000\n",
      "[2020-04-07 12:22:39] INFO: Selected k-mer size: 15\n",
      "[2020-04-07 12:22:39] INFO: >>>STAGE: assembly\n",
      "[2020-04-07 12:22:39] INFO: Assembling disjointigs\n",
      "[2020-04-07 12:22:39] INFO: Reading sequences\n",
      "[2020-04-07 12:22:40] INFO: Generating solid k-mer index\n",
      "[2020-04-07 12:22:55] INFO: Counting k-mers (1/2):\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:23:08] INFO: Counting k-mers (2/2):\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:23:33] INFO: Filling index table\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:24:02] INFO: Extending reads\n",
      "[2020-04-07 12:24:08] INFO: Overlap-based coverage: 27\n",
      "[2020-04-07 12:24:08] INFO: Median overlap divergence: 0.187975\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:25:03] INFO: Assembled 5 disjointigs\n",
      "[2020-04-07 12:25:03] INFO: Generating sequence\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:26:22] INFO: >>>STAGE: consensus\n",
      "[2020-04-07 12:26:22] INFO: Running Minimap2\n",
      "[2020-04-07 12:28:23] INFO: Computing consensus\n",
      "[2020-04-07 12:31:10] INFO: Alignment error rate: 0.197053\n",
      "[2020-04-07 12:31:10] INFO: >>>STAGE: repeat\n",
      "[2020-04-07 12:31:10] INFO: Building and resolving repeat graph\n",
      "[2020-04-07 12:31:10] INFO: Parsing disjointigs\n",
      "[2020-04-07 12:31:10] INFO: Building repeat graph\n",
      "10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:31:34] INFO: Median overlap divergence: 0.0367519\n",
      "[2020-04-07 12:31:35] INFO: Parsing reads\n",
      "[2020-04-07 12:31:36] INFO: Aligning reads to the graph\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:32:39] INFO: Aligned read sequence: 290447724 / 392320185 (0.740333)\n",
      "[2020-04-07 12:32:39] INFO: Median overlap divergence: 0.113942\n",
      "[2020-04-07 12:32:39] INFO: Mean edge coverage: 59\n",
      "[2020-04-07 12:32:39] INFO: Simplifying the graph\n",
      "[2020-04-07 12:32:40] INFO: >>>STAGE: contigger\n",
      "[2020-04-07 12:32:40] INFO: Generating contigs\n",
      "[2020-04-07 12:32:40] INFO: Reading sequences\n",
      "[2020-04-07 12:32:42] INFO: Generated 3 contigs\n",
      "[2020-04-07 12:32:42] INFO: Added 0 scaffold connections\n",
      "[2020-04-07 12:32:42] INFO: >>>STAGE: polishing\n",
      "[2020-04-07 12:32:42] INFO: Polishing genome (1/1)\n",
      "[2020-04-07 12:32:42] INFO: Running minimap2\n",
      "[2020-04-07 12:34:12] INFO: Separating alignment into bubbles\n",
      "[2020-04-07 12:37:21] INFO: Alignment error rate: 0.135813\n",
      "[2020-04-07 12:37:21] INFO: Correcting bubbles\n",
      "0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100% \n",
      "[2020-04-07 12:39:22] INFO: >>>STAGE: finalize\n",
      "[2020-04-07 12:39:22] INFO: Assembly statistics:\n",
      "\n",
      "\tTotal length:\t4642314\n",
      "\tFragments:\t2\n",
      "\tFragments N50:\t4640510\n",
      "\tLargest frg:\t4640510\n",
      "\tScaffolds:\t0\n",
      "\tMean coverage:\t79\n",
      "\n",
      "[2020-04-07 12:39:22] INFO: Final assembly: /home/toharhymes/work/ngs_ib_spring2020/project_6_evaluaton/assembling_cov80/assembly.fasta\n"
     ]
    }
   ],
   "source": [
    "!flye --pacbio-raw data/pacbio_80x.fq --out-dir assembling_cov80 --threads 8 -g 4.7m"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Сборка коротких ридов\n",
    "\n",
    "Соберите данные `Illumina` любым понравившимся вам сборщиком.\n",
    "\n",
    "Для пунктов 2 и 3 уду использовать spades.\n",
    "\n",
    "`spades -1 data/illumina.100x.1.fq -2 data/illumina.100x.2.fq -o \"illumina_assembling\"`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Гибридная сборка \n",
    "\n",
    "Сделайте гибридную сборку `Illumina` и `PacBio` (с разным покрытием) любым понравившимся вам сборщиком.\n",
    "\n",
    "`spades --pacbio data/pacbio_10x.fq -1 data/illumina.100x.1.fq -2 data/illumina.100x.2.fq -o \"hybrid_assembling_cov10\"`\n",
    "\n",
    "`spades --pacbio data/pacbio_20x.fq -1 data/illumina.100x.1.fq -2 data/illumina.100x.2.fq -o \"hybrid_assembling_cov20\"`\n",
    "\n",
    "`spades --pacbio data/pacbio_40x.fq -1 data/illumina.100x.1.fq -2 data/illumina.100x.2.fq -o \"hybrid_assembling_cov40\"`\n",
    "\n",
    "`spades --pacbio data/pacbio_80x.fq -1 data/illumina.100x.1.fq -2 data/illumina.100x.2.fq -o \"hybrid_assembling_cov80\"`\n",
    "\\# Почему то на последнем выдает ошибку, его не буду учитывать((("
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Cравнение сборок\n",
    "Сравните все получившиеся сборки с помощью [`QUAST`](http://quast.sourceforge.net/docs/manual.html) и данного референсного генома; интерпретируйте результаты. \n",
    "\n",
    "\n",
    "\n",
    "Во первых, можно сразу посмотреть на статистики из flye (в конце показывают):\n",
    "* Для покрытия 10:\n",
    "\n",
    "```\n",
    "Total length:\t4336541\n",
    "Fragments:\t97\n",
    "Fragments N50:\t60526\n",
    "Largest frg:\t155636\n",
    "Scaffolds:\t0\n",
    "Mean coverage:\t9\n",
    "\n",
    "```\n",
    "* Для покрытия 20:\n",
    "\n",
    "```\n",
    "Total length:\t4683515\n",
    "Fragments:\t26\n",
    "Fragments N50:\t314698\n",
    "Largest frg:\t718724\n",
    "Scaffolds:\t0\n",
    "Mean coverage:\t19\n",
    "```\n",
    "* Для покрытия 40:\n",
    "\n",
    "```\n",
    "Total length:\t4642057\n",
    "Fragments:\t1\n",
    "Fragments N50:\t4642057\n",
    "Largest frg:\t4642057\n",
    "Scaffolds:\t0\n",
    "Mean coverage:\t39\n",
    "```\n",
    "* Для покрытия 80:\n",
    "\n",
    "```\n",
    "Total length:\t4642314\n",
    "Fragments:\t2\n",
    "Fragments N50:\t4640510\n",
    "Largest frg:\t4640510\n",
    "Scaffolds:\t0\n",
    "Mean coverage:\t79\n",
    "```\n",
    "\n",
    "Сразу напрашивающийся вывод: уже при покрытии 40 ему удалось собрать все в 1 фрагмент. Круто! А покрытие 80 дало два фрагмента, но второй совсем маленький: меньше 2к нуклеотидов.\n",
    "\n",
    "Теперь полезем в quast. Файлы с выравниваниями:\n",
    "```\n",
    "assembling_cov10/assembly.fasta\n",
    "assembling_cov20/assembly.fasta\n",
    "assembling_cov40/assembly.fasta\n",
    "assembling_cov80/assembly.fasta\n",
    "\n",
    "illumina_assembling/scaffolds.fasta\n",
    "\n",
    "hybrid_assembling_cov10/scaffolds.fasta\n",
    "hybrid_assembling_cov20/scaffolds.fasta\n",
    "hybrid_assembling_cov40/scaffolds.fasta\n",
    "hybrid_assembling_cov80/scaffolds.fasta\n",
    "\n",
    "```\n",
    "И референс:\n",
    "```\n",
    "data/reference.fasta\n",
    "```\n",
    "\n",
    "\n",
    "Тогда кваст запустим следующей командой:\n",
    "```\n",
    "quast assembling_cov10/assembly.fasta \\\n",
    "        assembling_cov20/assembly.fasta \\\n",
    "        assembling_cov40/assembly.fasta \\\n",
    "        assembling_cov80/assembly.fasta \\\n",
    "        illumina_assembling/contigs.fasta \\\n",
    "        hybrid_assembling_cov10/contigs.fasta \\\n",
    "        hybrid_assembling_cov20/contigs.fasta \\\n",
    "        hybrid_assembling_cov40/contigs.fasta \\\n",
    "        -r data/reference.fasta\n",
    "```\n",
    "\n",
    "Вот что вышло:\n",
    "\n",
    "Summary table:\n",
    "![](img/table.pmg)\n",
    "Contig alignment viewer:\n",
    "\n",
    "*Legend:*\n",
    "* GREEN - correct contigs\n",
    "* RED - misassembled contigs\n",
    "* GREY - unaligned contigs\n",
    "* WHITE - unaligned parts of contigswith alignments)\n",
    "![](img/aligned.png)\n",
    "Cumulative length\n",
    "![](img/len.png)\n",
    "Contig size viewer:\n",
    "![](img/aligned.png)\n",
    "GC:\n",
    "![](img/gc.png)\n",
    "Nx:\n",
    "![](img/Nx.png)\n",
    "NAx:\n",
    "![](img/NAx.png)\n",
    "NGx\n",
    "![](img/NGx.png)\n",
    "NGAx:\n",
    "![](img/NGAx.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Выводы:\n",
    "1. Очень странно, непонятно и интересно, почему тут много не выровнялось на референс. Выравнялись только сборка иллюмины прям хорошо, и сборка pacbio с покрытием 10 и половина с покрытием 20 (но тут контигов много)\n",
    "2. Как и следовало ожидать, чем больше покрытие, тем меньше контигов (исключение: для покрытия 40 в чистом pacbio получился 1 контиг, а для покрытия 80 - 2, но уже с иллюминой наоборот:)\n",
    "3. При гибридной сборке количество контигов уменьшается (вцелом тоже логично)\n",
    "\n",
    "НО НЕПОНЯТНО ПОЧЕМУ ВСЕ СЕРОЕ.\n",
    "4. Статистики: GC контент логично у всех достаточно хорошо совпал. Остальные статистики (NGx, NAx логично лучше у тех сборок, у которых покрытие больше, и лучше если собирать с иллюминой)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
